### PR TITLE
Impression listener

### PR DESCRIPTION
**Pull request by maksimf**
- [ ] add more specs

### CODE REVIEW

**Comment by maksimf** [comment]
@patricioe This PR is ready for review

File: lib/cache/routers/impression_router.rb [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+    def router_thread
+      Thread.new do
+        loop do
+          @listener.log(@queue.pop)
+        end
+      end
+    end
+
+    def random_interval(interval)
```
Comment (patricioe) [inline lib/cache/routers/impression_router.rb]: where is this used?

File: lib/cache/routers/impression_router.rb [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        )
+      end
+    end
+
+    private
+
+    def router_thread
+      Thread.new do
+        loop do
+          @listener.log(@queue.pop)
```
Comment (patricioe) [inline lib/cache/routers/impression_router.rb]: what happens is there are exceptions here?

Comment (patricioe) [inline lib/cache/routers/impression_router.rb]: also, does it block if there are no impressions?

File: lib/splitclient-rb/clients/split_client.rb [hunk 3]
```diff
@@ -120,6 +123,30 @@ def store_impression(split_name, matching_key, bucketing_key, treatment, store_i
       )
     end
 
+    def route_impression(split_name, matching_key, bucketing_key, treatment, attributes)
```
Comment (patricioe) [inline lib/splitclient-rb/clients/split_client.rb]: no labels here?

File: lib/splitclient-rb/split_config.rb [hunk 4]
```diff
@@ -251,6 +257,10 @@ def self.default_impressions_refresh_rate
       60
     end
 
+    def self.default_impression_listener_refresh_rate
```
Comment (patricioe) [inline lib/splitclient-rb/split_config.rb]: Please make the default to be immediate

File: lib/cache/routers/impression_router.rb (hunk 2 above)
Comment (maksimf) [inline lib/cache/routers/impression_router.rb]: > what happens is there are exceptions here?
thread dies, it doesn't affect the main thread anyhow
> also, does it block if there are no impressions?
when they are no impressions thread just sleeps

File: lib/cache/routers/impression_router.rb (hunk 1 above)
Comment (maksimf) [inline lib/cache/routers/impression_router.rb]: Nowhere, deleted

File: lib/splitclient-rb/clients/split_client.rb (hunk 3 above)
Comment (maksimf) [inline lib/splitclient-rb/clients/split_client.rb]: label is stored inside `treatment[:label]`

File: lib/splitclient-rb/split_config.rb (hunk 4 above)
Comment (maksimf) [inline lib/splitclient-rb/split_config.rb]: Done

File: lib/cache/routers/impression_router.rb (hunk 2 above)
Comment (patricioe) [inline lib/cache/routers/impression_router.rb]: let's not let the thread dies. What happens after it dies? does it get created again?

Comment (maksimf) [inline lib/cache/routers/impression_router.rb]: Nope, it just dies and that's it.

Do you want me to do the same thing we do with other threads? i.e. log all exceptions in the infinite loop?

Comment (patricioe) [inline lib/cache/routers/impression_router.rb]: Yeah. If the thread dies it represents a problem for us.

Log should be warn.

File: CHANGES.txt [hunk 5]
```diff
@@ -1,3 +1,6 @@
+4.3.0 (July 12, 2017)
+- Add impression listener
+
```
Comment (patricioe) [inline CHANGES.txt]: Let's rebase this branch with latest develpment

File: lib/splitclient-rb/cache/routers/impression_router.rb [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+        )
+      end
+    end
+
+    private
+
+    def router_thread
+      Thread.new do
+        loop do
+          @listener.log(@queue.pop)
```
Comment (patricioe) [inline lib/splitclient-rb/cache/routers/impression_router.rb]: what happens if there is an exception here? will the thread die?

File: lib/splitclient-rb/cache/routers/impression_router.rb [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+            change_number: impressions[:treatments_labels_change_numbers][split_name.to_sym][:change_number]
+          },
+          attributes: impressions[:attributes]
+        )
+      end
+    end
+
+    private
+
+    def router_thread
```
Comment (patricioe) [inline lib/splitclient-rb/cache/routers/impression_router.rb]: what happens if there are no listeners? Can that happen?

File: lib/splitclient-rb/clients/split_client.rb [hunk 8]
```diff
@@ -116,7 +112,9 @@ def get_treatment(
       parsed_treatment(multiple, treatment_data)
     end
 
-    def store_impression(split_name, matching_key, bucketing_key, treatment, store_impressions)
+    def store_impression(split_name, matching_key, bucketing_key, treatment, store_impressions, attributes)
+      route_impression(split_name, matching_key, bucketing_key, treatment, attributes) if @config.impression_listener && store_impressions
+
       return if @config.impressions_queue_size <= 0 || !store_impressions
 
       @impressions_repository.add(split_name,
```
Comment (patricioe) [inline lib/splitclient-rb/clients/split_client.rb]: impressions are stored even if the listener is not set , correct?

File: spec/cache/routers/impression_router_spec.rb [hunk 9]
```diff
@@ -0,0 +1,31 @@
+require 'spec_helper'
+
+describe SplitIoClient::ImpressionRouter do
+  let(:dbl) { double }
```
Comment (patricioe) [inline spec/cache/routers/impression_router_spec.rb]: what is double?

Comment (maksimf) [inline spec/cache/routers/impression_router_spec.rb]: It's a dummy object which accepts any method on it, useful in tests

File: lib/splitclient-rb/cache/routers/impression_router.rb (hunk 7 above)
Comment (maksimf) [inline lib/splitclient-rb/cache/routers/impression_router.rb]: If there are no listeners, then we don't get here, see: https://github.com/splitio/ruby-client/blob/featute/impression_listener/lib/splitclient-rb/clients/split_client.rb#L116

`if @config.impression_listener && store_impressions`

File: lib/splitclient-rb/clients/split_client.rb (hunk 8 above)
Comment (maksimf) [inline lib/splitclient-rb/clients/split_client.rb]: yes, that's right

File: lib/splitclient-rb/cache/routers/impression_router.rb [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+    attr_reader :router_thread
+
+    def initialize(config)
+      @config = config
+      @listener = config.impression_listener
+      @queue = Queue.new
+      router_thread
+
+      if defined?(PhusionPassenger)
+        PhusionPassenger.on_event(:starting_worker_process) do |forked|
```
Comment (patricioe) [inline lib/splitclient-rb/cache/routers/impression_router.rb]: @maksimf do we need the same for unicorn and puma?

File: lib/splitclient-rb/cache/routers/impression_router.rb (hunk 6 above)
Comment (patricioe) [inline lib/splitclient-rb/cache/routers/impression_router.rb]: @maksimf did you address this issue? impressions listener thread should not die. Please use the same pattern as we use in the fetchers and regular impression sender.

File: lib/splitclient-rb/cache/routers/impression_router.rb (hunk 10 above)
Comment (maksimf) [inline lib/splitclient-rb/cache/routers/impression_router.rb]: No, for unicorn and puma we apply the fix from the Rails app side, there's also a ticket to switch to this approach for passenger: https://splitio.atlassian.net/browse/SPLIT-2604

File: lib/cache/routers/impression_router.rb (hunk 2 above)
Comment (maksimf) [inline lib/cache/routers/impression_router.rb]: Done

**Comment by maksimf** [comment]
@patricioe Done

File: lib/splitclient-rb/cache/routers/impression_router.rb (hunk 7 above)
Comment (patricioe) [inline lib/splitclient-rb/cache/routers/impression_router.rb]: you sure? the link seems outdated.

I see `route_impression` being called on every Store_impressions. Mind helping understand the flow?

```ruby
    def store_impression(split_name, matching_key, bucketing_key, treatment, store_impressions, attributes)
      route_impression(split_name, matching_key, bucketing_key, treatment, attributes) if @config.impression_listener && store_impressions

      return if @config.impressions_queue_size <= 0 || !store_impressions

      @impressions_repository.add(split_name,
        'keyName' => matching_key,
        'bucketingKey' => bucketing_key,
        'treatment' => treatment[:treatment],
        'label' => @config.labels_enabled ? treatment[:label] : nil,
        'time' => (Time.now.to_f * 1000.0).to_i,
        'changeNumber' => treatment[:change_number]
      )
    end
```

Comment (patricioe) [inline lib/splitclient-rb/cache/routers/impression_router.rb]: > what happens if there are no custom listeners?

Comment (maksimf) [inline lib/splitclient-rb/cache/routers/impression_router.rb]: See the end of the line, I pointed out in the previous comment:
`if @config.impression_listener && store_impressions`
If `@config.impression_listener` is `nil`, i.e impression_listener is not set `route_impression` is not called

Comment (maksimf) [inline lib/splitclient-rb/cache/routers/impression_router.rb]: In ruby you can wrirte something like this:
```ruby
do_this if that
```
Which is equal to:
```ruby
if that
  do_this
end
```

Comment (maksimf) [inline lib/splitclient-rb/cache/routers/impression_router.rb]: So, the part of the code you mentioned is equal to:
```ruby
    def store_impression(split_name, matching_key, bucketing_key, treatment, store_impressions, attributes)
      if @config.impression_listener && store_impressions
        route_impression(split_name, matching_key, bucketing_key, treatment, attributes) 
      end

      return if @config.impressions_queue_size <= 0 || !store_impressions

      @impressions_repository.add(split_name,
        'keyName' => matching_key,
        'bucketingKey' => bucketing_key,
        'treatment' => treatment[:treatment],
        'label' => @config.labels_enabled ? treatment[:label] : nil,
        'time' => (Time.now.to_f * 1000.0).to_i,
        'changeNumber' => treatment[:change_number]
      )
    end
```

Comment (patricioe) [inline lib/splitclient-rb/cache/routers/impression_router.rb]: my bad. I didn't see the end of the line.

**Pull request merged**